### ISSUE TITLE
Add support for HDR images 

### ISSUE THREAD

**Issue by Mattis142** [bluesky-social/social-app#6172]
### Describe the Feature

I'd like to see support for HDR images in Bluesky. I guess ideally realized through gainmaps/Google Ultra HDR style so that the same image file would still retain good colors on an SDR Display too. Instagram also added support for HDR images by utilizing Ultra HDR

### Attachments

Documentation
https://developer.android.com/media/platform/hdr-image-format

### Describe Alternatives

avif I guess?

**Comment by djhindley** [comment]
Yes, I agree. This would be a great feature. Given that most devices which are used by Bluesky users are highly likely to be capable of showing HDR images already, it would seem sensible to show the benefit of HDR in Bluesky image posts.

**Comment by digi2life** [comment]
Google and Apple have adopted the ISO 21496-1 standard for jpeg with HDR gain maps. That's definitely the way to implement this. Was hoping to also see better universal support for avif but but we're just not there yet.

**Comment by Mattis142** [comment]
especially since both Google and Apple seem to save HDR photos as JPEGs now, I know bluesky adds compression to images but I'm sure you can also compress a gainmap too using Google's ultra HDR library? https://github.com/google/libultrahdr tbf I'm not a dev but it seems like something that's possible

**Comment by gregbenz** [comment]
Photos with a gain map are the way to go. AVIF is an excellent format that does support HDR, but it needs a **gain map** in order to produce high quality results when showing an HDR image on a display which is SDR or has only limited HDR. Most phones have great HDR support, but many images will be encoded up to 4 stops of HDR (while phones typically offer 1.5 to 3 stops of headroom). So a simple HDR image is not nearly as good as an image with a gain map. In the future, AVIF with a gain map should be a viable option (and will offer both smaller files and higher quality than JPG with a gain map).

**JPG with a gain map** is the way to go now. ISO JPG gain maps look great on any display, are 100% backwards compatible, and are supported by MacOS, iOS, iPadOS, Android, Chrome, Edge, Brave, Opera, Adobe Camera RAW, Lightroom, etc. 

The Google-backed, open-source **libultrahdr** library may be used to transcode uploaded images, it provides excellent results. ImageMagick has also integrated that library, and there are similar requests to incorporate it into other libraries. And support for AVIF and HEIC gain maps is planned for libultra, making this library an excellent choice for multiple reasons.

Is BlueSky's current transcoding based on an open-source library such as ImageMagick?

More info on HDR gain maps: https://gregbenzphotography.com/hdr/#developers

**Comment by gregbenz** [comment]
[Safari Tech Preview 216](https://developer.apple.com/documentation/safari-technology-preview-release-notes/stp-release-216) has added support for HDR images. HDR photos are quickly becoming a mainstream technology. Almost universally supported on Apple products, most Android phones, and a quickly growing number of monitors and PC laptops. 80%+ of viewers already have support to appreciate the benefit of HDR on Instagram.

**Comment by clshortfuse** [comment]
We use HDR pngs in Discord and they work well in Chrome. I wanted to showcase some HDR stuff i've been working on, but that can't be done with Bluesky.

SDR:

![Image](https://github.com/user-attachments/assets/a4979b66-163b-4dd6-9562-c1d3f71c2763)

HDR:

![Image](https://github.com/user-attachments/assets/1b3da12e-4cbe-4775-a0f0-7e12f2830590)

Seems to work well here in GitHub as well.

If you can view this yellow orb in the HDR png, then you're viewing HDR:

![Image](https://github.com/user-attachments/assets/e5e9d832-65c2-4969-a0fe-310622c4a59a)

**Comment by gregbenz** [comment]
@clshortfuse That's interesting and good to know on Discord! That means support will work in Brave, Edge, and Opera too. I just tried and see HDR support in the MacOS Discord app (using PNG with a CICP tag, which you can export from Adobe Lightroom / Adobe Camera RAW). 

I tried Safari Tech Preview 220 (which supports PQ encoded JPG, AVIF, JXL) and did not see HDR there. But it's an unusual format and Safari support is just in the early preview stages. MacOS Preview supports it, so hopefully Safari will too soon. Great to see HDR on Discord.

**Comment by clshortfuse** [comment]
If I'm not mistaken, AVIF is currently the most universally supported format. Ideally speaking an HDR PNG could be transcoded by the server into the compatible AVIF format for wider compatibility.

Edit: Some samples from https://github.com/AOMediaCodec/av1-avif/tree/master/testFiles/Netflix/avif

![Image](https://raw.githubusercontent.com/AOMediaCodec/av1-avif/refs/heads/master/testFiles/Netflix/avif/hdr_cosmos01650_cicp9-16-9_yuv420_limited_qp10.avif)

**Comment by gregbenz** [comment]
@clshortfuse AVIF is ideal for HDR in the not too distant future (ie 2026). Support is ~95% of browsers by usage, which is less than PNG. However, support for HDR encoding of AVIF is very high, all modern browsers support AVIF (with HDR support in Chrome, Edge, Brave, Opera, and in Safari Tech Preview - FireFox doesn't support HDR photos in any format), and the images are a fraction of the size of a similar PNG. Support for **AVIF with a gain map** (which is critical to ensure highest image quality even on less capable displays) is coming soon (available under a dev flag in Chrome).

The most compatible format for HDR today is a **JPG with a gain map**. It is 100% safe / backwards compatible and support to show the HDR version is very high.

So support for both formats would be ideal: JPG as the priority to eliminate edge cases today with legacy software, and HDR AVIF with a gain map as soon as possible (based on rates of adoption) because it is a much better format (much smaller files and higher quality).

An AVIF gain map can be encoded with either a base HDR and a map to derive the SDR, or an SDR base with a map to derive the HDR (JPG only uses an SDR base due to limitations of an 8-bit file format). Ideally, either map approach would be supported - but a base HDR AVIF with a map to generate an SDR is most important. We already have more support for HDR AVIF than AVIF gain maps, so this offers more HDR benefit. But a map is required to ensure high quality (tone mapping of HDR down to lesser displays is often terrible) and an image where the map is just used to derive the SDR can be compressed more aggressively. So an **AVIF gain map with HDR as the base image offers greatest HDR benefit, highest quality, and the smallest files**.

**Comment by Mattis142** [comment]
I also would prefer JPEG with a gain map, simply because downloading it wouldn't require conversion to a different format as downloading an AVIF file often causes a lot of incompatibly across apps like different OEMs gallery apps, older software versions, users who want to download an image from Bluesky and then post it to a chat client of their choice that might not support AVIF yada yada

**Comment by clshortfuse** [comment]
@gregbenz I just realized who I was talking to 😅 

I'm personally trying to figure out best compatibility. But it does seem you've done the research on the JPG support. I'm sure you're in the same boat as I am with a great amount of HDR images that can't be easily shared.

I'm assuming it's this spec:

https://helpx.adobe.com/content/dam/help/en/camera-raw/using/gain-map/jcr_content/root/content/flex/items/position/position-par/table/row-3u03dx0-column-4a63daf/download_section/download-1/Gain_Map_1_0d15.pdf

Spec seems fine. I'm sure there's some caveats with colorspace.

It's worth sharing that Chrome just broken HDR PNG recently in some unknown way: https://issues.chromium.org/issues/423310049

But the alternative is for Bluesky not just "support HDR images", but just don't break it. If there's no reason for Bluesky to strip the JPEG gain map, or the AVIF gain map, just don't. Thumbnails and the like can be reencoded, but we should be able to see the unmodified image. This is actually how Discord works. Previews are re-encoded and sometimes in SDR, but once you open up the full preview, it's HDR (unmodified). Also, that's less stress on the server in terms of CPU.

Of course HDR PNGs can be massive (20MB), so there being some sort of known limit would be very useful.

**Comment by gregbenz** [comment]
@clshortfuse Agreed, "don't break it" is a good way of putting things in general terms. Put another way, any transcoding BlueSky does for a supported MIME type like JPG should be visually consistent with the course (but might be cropped, resized, or compressed). In practice, it's more complicated as there are wide range of possible encodings and BlueSky will need to focus on supporting transcoding for formats which really matter. I believe that means support for ISO encoded gain maps in JPG and AVIF file formats (with support for CICP in AVIF - PQ encoding in a full ICC profile is possible, but increases file size and reduces compatibility without any benefit for HDR on the web).

That Adobe spec and other approaches using XMP to encode metadata (including a published Android spec and Apple's propritary approach) are effectively legacy formats now. The target for any support now should be compliant with ISO 21496-1 (gain maps), ISO 22028-5 (HDR metadata) and as appropriate ISO 23091-2 (CICP encoding of color information). I have more info at https://gregbenzphotography.com/hdr/#developers (including links to open source libraries which support transcoding).

There are other viable HDR gain map formats like HEIF and JXL which have some support, but I see no reason to spend developer resources on serving them at this time as browser support is limited to Safari (transcoding HEIF to another supported gain map format is important to support iPhone uploads). I would love to see JXL become well supported, but there is no indication that is coming any time soon. Odds are even lower for HEIF. JPG is the best option today, and AVIF will soon be a better option.

**Comment by surfdude29** [comment]
> […] I would love to see JXL become well supported, but there is no indication that is coming any time soon. […]

Development is certainly coming along on [jxl-rs](https://github.com/libjxl/jxl-rs), a "reimplementation of a JPEG XL decoder in Rust, which aims to be conforming, safe and fast", which will allow Firefox (and Chromium? 🤞) to finally implement JXL support.

**Comment by surfdude29** [comment]
Also, [just announced today](https://webkit.org/blog/16993/news-from-wwdc25-web-technology-coming-this-fall-in-safari-26-beta/#hdr-images):

> Now, in the Safari 26 beta, WebKit adds support for HDR images on the web. You can embed images with high dynamic range into a webpage, just like other images — including images in Canvas.

**Comment by clshortfuse** [comment]
I had uploaded a 300kb AVIF with HDR and Bluesky converted it to a 425KB SDR-only JPEG, which seems counter-productive.

I used [libavif](https://github.com/AOMediaCodec/libavif/releases) to encode down from a 3.8MB png. 

I had tried Google's [libultrahdr](https://github.com/google/libultrahdr/), but it doesn't support PNG or AVIF as input, really complicating the task. You would also need to manually type in the width and height.

Unfortunately libavif doesn't seem to support HDR (Gain Map) JPG => AVIF without dropping it down to SDR, so it's not a silver bullet.

**Comment by gregbenz** [comment]
@clshortfuse ImageMagick has integrated libultrahdr and can be used to support more inputs (need to enable support when you compile). https://github.com/ImageMagick/ImageMagick/issues/6377
